File: Problem Set 1/credit/credit.c
// credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdbool.h>

static int luhn_checksum(long long n)
{
    int sum = 0;
    bool double_next = false;

    while (n > 0)
    {
        int d = n % 10;
        if (double_next)
        {
            int x = d * 2;
            sum += (x / 10) + (x % 10);
        }
        else
        {
            sum += d;
        }
        double_next = !double_next;
        n /= 10;
    }
    return sum % 10;
}

static int length_of(long long n)
{
    int len = 0;
    while (n > 0)
    {
        len++;
        n /= 10;
    }
    return len;
}

int main(void)
{
    long long number = get_long("Number: ");

    if (number <= 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Keep originals for checks
    long long n = number;

    // Compute length
    int len = length_of(n);

    // Grab first two digits and first digit
    long long tmp = n;
    while (tmp >= 100)
    {
        tmp /= 10;
    }
    int first_two = (int) tmp;
    int first_one = first_two / 10;

    // Luhn validity
    if (luhn_checksum(number) != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Identify issuer
    if (len == 15 && (first_two == 34 || first_two == 37))
    {
        printf("AMEX\n");
    }
    else if (len == 16 && (first_two >= 51 && first_two <= 55)) // classic MasterCard range for this problem
    {
        printf("MASTERCARD\n");
    }
    else if ((len == 13 || len == 16) && first_one == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

File: Problem Set 6/credit/credit.py
def luhn_checksum(n: int) -> int:
    total = 0
    double_next = False
    while n > 0:
        d = n % 10
        if double_next:
            x = d * 2
            total += x // 10 + x % 10
        else:
            total += d
        double_next = not double_next
        n //= 10
    return total % 10

def length_of(n: int) -> int:
    length = 0
    while n > 0:
        length += 1
        n //= 10
    return length

def main():
    while True:
        try:
            number = int(input("Number: "))
            break
        except ValueError:
            continue

    if number <= 0:
        print("INVALID")
        return

    n = number
    length = length_of(n)

    tmp = n
    while tmp >= 100:
        tmp //= 10
    first_two = int(tmp)
    first_one = first_two // 10

    if luhn_checksum(number) != 0:
        print("INVALID")
        return

    if length == 15 and (first_two == 34 or first_two == 37):
        print("AMEX")
    elif length == 16 and (51 <= first_two <= 55):
        print("MASTERCARD")
    elif (length == 13 or length == 16) and first_one == 4:
        print("VISA")
    else:
        print("INVALID")

if __name__ == "__main__":
    main()
